imap/jmap_mail.c

```diff
@@ -13236,6 +13236,63 @@ static int msgimport_checkacl_cb(const mbentry_t *mbentry, void *xrock)
     return 0;
 }
 
+static time_t _email_import_parse_received_at(const char *blob, size_t blob_len)
+{
+    message_t *msg = message_new_from_data(blob, blob_len);
+    struct buf buf = BUF_INITIALIZER;
+    time_t received_at = 0;
+    enum message_format format =  MESSAGE_DECODED | MESSAGE_TRIM;
+
+    if (!msg)
+        goto done;
+
+    if (!message_get_field(msg, "X-Deliveredinternaldate", format, &buf)) {
+        buf_appendcstr(&buf, "\r\n");
+        const char *hdr = buf_cstring(&buf);
+        char *val = NULL;
+        message_parse_string(hdr, &val);
+        if (time_from_rfc822(val, &received_at) < 0)
+            received_at = 0;
+        xzfree(val);
+    }
+
+    if (received_at)
+        goto done;
+
+    if (!message_get_field(msg, "Received", format, &buf)) {
+        buf_appendcstr(&buf, "\r\n");
+        const char *hdr = buf_cstring(&buf);
+        char *val = NULL;
+        do {
+            message_parse_received_date(hdr, &val);
+            if (time_from_rfc822(val, &received_at) < 0)
+                received_at = 0;
+            hdr = strchr(hdr, '\n');
+            if (hdr)
+                hdr = strchr(hdr + 1, ':');
+            xzfree(val);
+        } while (!received_at && hdr++);
+    }
+
+    if (received_at)
+        goto done;
+
```

Comment:
Same comment as above regarding goto.

Also, since parsing X-Deliveredinternaldate and Date are the same, I wonder if we could factor this out into a separate function that accept the header name as an argument.

And if we decide to only check the first Received header, we *may* be able to roll that into the new function as well.

Reply:
We will always want to check all Received headers, since otherwise we could use the logic from message.c.

I figured it's not worth it to refactor the Date and X-Deliveredinternaldate code into a function since it's just two occasions where we need it. But I can refactor if you prefer.

Reply:
OK, that's fine.  But we should change the description to state that we check all Received headers.

Reply:
Which description?

Reply:
From the PR description:

This updates Email/import to set the receivedAt property to:

    the Email/import receivedAt argument, if present
    the value of the X-Deliveredinternaldate header, if present
    the value of the **first** valid Received header, if present

Reply:
I rephrased this to
```
- the value of the first valid Received header, of all Received headers
```
to make this more clear